File: app/src/main/java/com/touchnothing/app/domain/AuthErrorMessage.kt
package com.touchnothing.app.domain

import android.content.Context
import com.touchnothing.app.R
import com.touchnothing.app.data.model.AuthValidationError
import com.touchnothing.app.data.remote.SupabaseServiceError
import com.touchnothing.app.util.NicknameValidationError

object AuthErrorMessage {
    fun validationMessage(context: Context, error: AuthValidationError): String =
        when (error) {
            is AuthValidationError.Nickname -> nicknameMessage(context, error.error)
            AuthValidationError.InvalidPin -> context.getString(R.string.auth_invalid_pin)
            AuthValidationError.PinMismatch -> context.getString(R.string.auth_pin_mismatch)
        }

    fun nicknameMessage(context: Context, error: NicknameValidationError): String =
        when (error) {
            NicknameValidationError.EMPTY -> context.getString(R.string.nickname_empty)
            NicknameValidationError.INVALID_LENGTH -> context.getString(R.string.nickname_invalid_length)
            NicknameValidationError.INVALID_CHARACTERS -> context.getString(R.string.nickname_invalid_characters)
        }

    fun serviceMessage(context: Context, error: SupabaseServiceError): String =
        when (error) {
            SupabaseServiceError.NicknameTaken -> context.getString(R.string.nickname_taken)
            is SupabaseServiceError.InvalidCredentials -> {
                error.remainingAttempts?.let { remaining ->
                    context.getString(R.string.auth_invalid_credentials_remaining, remaining)
                } ?: context.getString(R.string.auth_invalid_credentials)
            }
            SupabaseServiceError.AccountLocked -> context.getString(R.string.auth_account_locked)
            SupabaseServiceError.InvalidPin -> context.getString(R.string.auth_invalid_pin)
            SupabaseServiceError.InvalidNickname -> context.getString(R.string.nickname_invalid_length)
            SupabaseServiceError.InvalidSessionToken -> context.getString(R.string.auth_session_expired)
            SupabaseServiceError.NicknameNotRegistered -> context.getString(R.string.auth_invalid_credentials)
            SupabaseServiceError.ServerUpgradeRequired -> context.getString(R.string.auth_server_upgrade_required)
            SupabaseServiceError.NotConfigured,
            SupabaseServiceError.NetworkFailure,
            -> context.getString(R.string.nickname_network_error)
        }
}

File: TouchNothing/Utilities/AuthErrorMessage.swift
import Foundation

enum AuthErrorMessage {
    static func message(for validationError: AuthValidationError) -> String {
        switch validationError {
        case .nickname(let nicknameError):
            return message(for: nicknameError)
        case .invalidPin:
            return LocalizationKey.authInvalidPin.localized
        case .pinMismatch:
            return LocalizationKey.authPinMismatch.localized
        }
    }

    static func message(for validationError: NicknameValidationError) -> String {
        switch validationError {
        case .empty:
            return LocalizationKey.nicknameEmpty.localized
        case .invalidLength:
            return LocalizationKey.nicknameInvalidLength.localized
        case .invalidCharacters:
            return LocalizationKey.nicknameInvalidCharacters.localized
        }
    }

    static func message(for serviceError: SupabaseServiceError) -> String {
        switch serviceError {
        case .nicknameTaken:
            return LocalizationKey.nicknameTaken.localized
        case .invalidCredentials:
            return LocalizationKey.authInvalidCredentials.localized
        case .accountLocked:
            return LocalizationKey.authAccountLocked.localized
        case .invalidPin:
            return LocalizationKey.authInvalidPin.localized
        case .invalidNickname:
            return LocalizationKey.nicknameInvalidLength.localized
        case .invalidSessionToken:
            return LocalizationKey.authSessionExpired.localized
        case .nicknameNotRegistered:
            return LocalizationKey.authInvalidCredentials.localized
        case .serverUpgradeRequired:
            return LocalizationKey.authServerUpgradeRequired.localized
        case .notConfigured, .networkFailure:
            return LocalizationKey.nicknameNetworkError.localized
        }
    }
}
